### PR TITLE
Enip rust 3958 v9

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3958

Describe changes:
- convert enip parser to rust
- integer keywords now support hexadecimal notation

Alon the way, also
- transactions are now bidirectional
- there is a enip logger
- gap support is improved with probing for resync
- frames
- events
- enip_command keyword accepts now string enumeration as values.
- more keywords

#10048 with needed rebase (DetectAppLayerMpmRegister2 renamed to DetectAppLayerMpmRegister) and auto squash

### Provide values to any of the below to override the defaults.

```
SV_BRANCH=pr/1521
```
https://github.com/OISF/suricata-verify/pull/1521

Does the first commit deserve its own redmine ticket ?
And the one in https://github.com/OISF/suricata/pull/9945/commits/4a493528fa694e50a5f0bbd1f92167e2ba9241e7 also ?

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 181 | 192 | 106.08% | 

Pipeline 17157

File: rust/src/enip/detect.rs [hunk 1]
```diff
@@ -0,0 +1,808 @@
...
+fn enip_parse_command(i: &str) -> IResult<&str, u16> {
+    let (i, v) = alt((enip_detect_parse_u16, enip_parse_command_string))(i)?;
+    return Ok((i, v));
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_parse_command(
+    raw: *const std::os::raw::c_char, value: *mut u16,
+) -> bool {
+    let raw2: &CStr = CStr::from_ptr(raw); //unsafe
```
Comment (jasonish) [inline rust/src/enip/detect.rs:75]: Nothing wrong here, but in Rust its pretty normal to shadow the previous variable.

File: rust/src/enip/detect.rs [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+                    return true;
+                }
+            }
+        }
+    }
+    return false;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_get_serial(
```
Comment (jasonish) [inline rust/src/enip/detect.rs:612]: I think we should start naming these like C functions..  ie) https://github.com/OISF/suricata/blob/master/rust/src/dns/dns.rs#L883

File: rust/src/enip/detect.rs [hunk 3]
```diff
@@ -0,0 +1,808 @@
...
+    tx: *mut std::os::raw::c_void, value: *mut u32,
```
Comment (jasonish) [inline rust/src/enip/detect.rs:613]: If we're not checking these for NULL, we can get rid of the cast by `tx: &mut EnipTransaction`.  See: https://github.com/OISF/suricata/blob/master/rust/src/dns/dns.rs#L884

File: rust/src/enip/detect.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/enip/detect.rs:75]: nice

File: rust/src/enip/detect.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/enip/detect.rs:613]: indeed

File: rust/src/enip/detect.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/enip/detect.rs:612]: done

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/10083 thank you @jasonish

**Pull request closed without merging**